### PR TITLE
EAHW-2577: Create person producer

### PR DESCRIPTION
**Pull request by joannafawl**
- Created PersonKafkaEntityListener which sends Kafka message when creating/updating/deleting a person
- Added integration tests

### CODE REVIEW

File: src/test/java/uk/gov/homeoffice/digital/sas/person/CallistoPersonRestApiIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.MvcResult;
+import org.springframework.test.web.servlet.ResultActions;
+import uk.gov.homeoffice.digital.sas.kafka.message.KafkaAction;
+import uk.gov.homeoffice.digital.sas.kafka.message.KafkaEventMessage;
+import uk.gov.homeoffice.digital.sas.person.enums.TermsAndConditions;
+import uk.gov.homeoffice.digital.sas.person.model.Person;
+import uk.gov.homeoffice.digital.sas.person.testutils.KafkaConsumer;
+
+@RunWith(SpringRunner.class)
```
Comment (daleelR) [inline src/test/java/uk/gov/homeoffice/digital/sas/person/CallistoPersonRestApiIntegrationTest.java]: I believe the junit 5 equivalent/recommendation is to use `@ExtendWith(SpringExtension.class) `
But I don't think you need it anyway since you have `@SpringBootTest `as mentioned in: 
https://docs.spring.io/spring-boot/docs/current/reference/html/features.html#features.testing.spring-boot-applications

File: src/test/java/uk/gov/homeoffice/digital/sas/person/CallistoPersonRestApiIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+  @Test
+  void shouldReturnEmptyListOfItemsWhenNoPersonDataPresent() throws Exception {
+
+    mvc.perform(get(PERSON_URL+ TENANT_ID_PARAM)
+            .contentType(MediaType.APPLICATION_JSON))
+        .andExpect(status().isOk())
+        .andExpect(content()
+            .contentTypeCompatibleWith(MediaType.APPLICATION_JSON))
+        .andExpect(jsonPath("$.items", empty()));
```
Comment (daleelR) [inline src/test/java/uk/gov/homeoffice/digital/sas/person/CallistoPersonRestApiIntegrationTest.java:95]: It feels like there should be an assertion before this one to assert that there actually isn't any person data in the db.
You might be able to just autowire in the repo and check it's empty.

File: src/test/java/uk/gov/homeoffice/digital/sas/person/listeners/PersonKafkaEntityListenerTest.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+  @BeforeEach
+  void setup() {
+    person = Person.builder()
+        .firstName("John")
+        .lastName("Smith")
+        .version(1)
+        .fteValue(new BigDecimal("1.0000"))
+        .termsAndConditions(TermsAndConditions.MODERNISED)
+        .build();
```
Comment (daleelR) [inline src/test/java/uk/gov/homeoffice/digital/sas/person/listeners/PersonKafkaEntityListenerTest.java:38]: nit: you could probs get away with not having to set any of these fields.
No harm but doesn't seem relevant to the test.

File: src/test/java/uk/gov/homeoffice/digital/sas/person/CallistoPersonRestApiIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+import uk.gov.homeoffice.digital.sas.kafka.message.KafkaEventMessage;
+import uk.gov.homeoffice.digital.sas.person.enums.TermsAndConditions;
+import uk.gov.homeoffice.digital.sas.person.model.Person;
+import uk.gov.homeoffice.digital.sas.person.testutils.KafkaConsumer;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.MOCK)
+@DirtiesContext(classMode = ClassMode.AFTER_EACH_TEST_METHOD)
+@AutoConfigureMockMvc
+@TestPropertySource(locations = "classpath:application.properties")
```
Comment (daleelR) [inline src/test/java/uk/gov/homeoffice/digital/sas/person/CallistoPersonRestApiIntegrationTest.java]: nit: I don't think you need this annotation either as the default property source with `@SpringBootTest` is `test/resources/application.properties`

File: src/test/java/uk/gov/homeoffice/digital/sas/person/CallistoPersonRestApiIntegrationTest.java (hunk 2 above)
Comment (taomoh) [reply src/test/java/uk/gov/homeoffice/digital/sas/person/CallistoPersonRestApiIntegrationTest.java:95]: I don't see the need for the additional assertion. H2 database should be reset before each test thanks to `@DirtiesContext(classMode = ClassMode.AFTER_EACH_TEST_METHOD)`

File: src/test/java/uk/gov/homeoffice/digital/sas/person/listeners/PersonKafkaEntityListenerTest.java (hunk 3 above)
Comment (taomoh) [reply src/test/java/uk/gov/homeoffice/digital/sas/person/listeners/PersonKafkaEntityListenerTest.java:38]: The only field that can be left out, without failing the test, is fteValue. All the others are mandatory (@NotNull).
Still, I feel it's sensible to include it for completeness

File: src/test/java/uk/gov/homeoffice/digital/sas/person/CallistoPersonRestApiIntegrationTest.java (hunk 1 above)
Comment (taomoh) [reply src/test/java/uk/gov/homeoffice/digital/sas/person/CallistoPersonRestApiIntegrationTest.java]: Well-spotted @daleelR . I have now removed these superfluous annotations

File: src/test/java/uk/gov/homeoffice/digital/sas/person/CallistoPersonRestApiIntegrationTest.java (hunk 4 above)
Comment (taomoh) [reply src/test/java/uk/gov/homeoffice/digital/sas/person/CallistoPersonRestApiIntegrationTest.java]: another good spot @daleelR . I have now removed it

File: src/test/java/uk/gov/homeoffice/digital/sas/person/CallistoPersonRestApiIntegrationTest.java (hunk 2 above)
Comment (daleelR) [reply src/test/java/uk/gov/homeoffice/digital/sas/person/CallistoPersonRestApiIntegrationTest.java:95]: Ah yes I forgot that annotation was present. You're right, in this context the DB would always be clear.

File: src/test/java/uk/gov/homeoffice/digital/sas/person/listeners/PersonKafkaEntityListenerTest.java (hunk 3 above)
Comment (daleelR) [reply src/test/java/uk/gov/homeoffice/digital/sas/person/listeners/PersonKafkaEntityListenerTest.java:38]: `@NotNull` would only take effect when invoking the validator. Here it's just a plain instantiation so should still pass.

My thinking leans towards keeping test setup to a minimum for what's required to make the test fulfill it's scenario and pass.
But that's only my personal preference and absolutely no harm to include. Will resolve 👍

File: src/main/java/uk/gov/homeoffice/digital/sas/person/listeners/PersonKafkaEntityListener.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    super(kafkaProducerService);
+  }
+
+  @Override
+  public String resolveMessageKey(Person person) {
+    return person.getTenantId() + ":" + person.getId();
+  }
+
+  @PostPersist
+  void sendMessageOnCreate(Person resource) {
```
Comment (stephenkaye-ho) [inline src/main/java/uk/gov/homeoffice/digital/sas/person/listeners/PersonKafkaEntityListener.java:24]: I wasn't expecting to see these methods being required, are they not inherited from the base class.
I thought only the resolveMessageKey would be abstract and need an implementation

File: src/test/java/uk/gov/homeoffice/digital/sas/person/CallistoPersonRestApiIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+
+  @Autowired
+  private ObjectMapper objectMapper;
+
+  @Autowired
+  private KafkaConsumer kafkaConsumer;
+
+  @BeforeEach
+  void setup() {
+    person = Person.builder()
```
Comment (stephenkaye-ho) [inline src/test/java/uk/gov/homeoffice/digital/sas/person/CallistoPersonRestApiIntegrationTest.java:78]: I think we should randomise the person's name so that each test is different. At the moment, when the message on the queue is verified, I can't tell if it's reading the same message all the time as the person object will always be equal.

**Review by stephenkaye-ho — DISMISSED** [review]

File: src/main/java/uk/gov/homeoffice/digital/sas/person/listeners/PersonKafkaEntityListener.java (hunk 5 above)
Comment (taomoh) [reply src/main/java/uk/gov/homeoffice/digital/sas/person/listeners/PersonKafkaEntityListener.java:24]: This is being addressed separately: https://github.com/UKHomeOffice/callisto-jparest/pull/58

File: src/test/java/uk/gov/homeoffice/digital/sas/person/CallistoPersonRestApiIntegrationTest.java (hunk 6 above)
Comment (taomoh) [reply src/test/java/uk/gov/homeoffice/digital/sas/person/CallistoPersonRestApiIntegrationTest.java:78]: Actioned

**Review by daleelR — APPROVED** [review]

**Review by claire-guest — APPROVED** [review]
